flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/sink/ConstraintValidator.java

```diff
@@ -0,0 +1,209 @@
+                        "ConstraintValidator requires that there are CHAR/VARCHAR fields.");
+                this.charFields = charFields;
+                this.allFieldNames = allFieldNames;
+                this.mustApply = true;
+            }
+        }
+
+        public ConstraintValidator build() {
+            return new ConstraintValidator(
+                    notNullEnforcer,
+                    notNullFieldIndices,
+                    charPrecisionEnforcer,
+                    charFields != null ? charFields.stream().mapToInt(t -> t.f0).toArray() : null,
+                    charFields != null ? charFields.stream().mapToInt(t -> t.f1).toArray() : null,
+                    allFieldNames);
+        }
+    }
+
+    @Override
+    public void processElement(StreamRecord<RowData> element) throws Exception {
+        StreamRecord<RowData> processedElement = processNotNullConstraint(element);
+        if (processedElement != null) {
+            output.collect(processCharConstraint(processedElement));
+        }
+    }
+
+    private StreamRecord<RowData> processNotNullConstraint(StreamRecord<RowData> element) {
+        if (notNullEnforcer == null) {
+            return element;
+        }
+
+        final RowData rowData = element.getValue();
+
+        for (int index : notNullFieldIndices) {
+            if (rowData.isNullAt(index)) {
+                switch (notNullEnforcer) {
+                    case ERROR:
+                        throw new TableException(
+                                String.format(
+                                        "Column '%s' is NOT NULL, however, a null value is being written into it. "
+                                                + "You can set job configuration '%s'='drop' "
+                                                + "to suppress this exception and drop such records silently.",
+                                        allFieldNames[index],
+                                        TABLE_EXEC_SINK_NOT_NULL_ENFORCER.key()));
+                    case DROP:
+                        return null;
+                }
+            }
+        }
+        return element;
+    }
+
+    private StreamRecord<RowData> processCharConstraint(StreamRecord<RowData> element) {
+        if (charPrecisionEnforcer == null
+                || charPrecisionEnforcer == CharPrecisionEnforcer.IGNORE) {
+            return element;
+        }
+
+        final RowData rowData = element.getValue();
+
+        boolean trimmed = false;
+        for (int i = 0; i < charFieldIndices.length; i++) {
+            final int fieldIdx = charFieldIndices[i];
+            final int precision = charFieldPrecisions[i];
+            final String stringValue = rowData.getString(fieldIdx).toString();
+
+            if (stringValue.length() > precision) {
+                if (!trimmed) {
+                    reusableRowData.setRow(rowData);
+                }
+                reusableRowData.setField(
```

Comment:
For performing the trimming, there is a function in `BinaryStringDataUtil` called `substringSQL` which should be optimized for trimming directly the underlying memory segment. It accepts `BinaryStringData` more than the `StringData` interface, but you can do an `instanceof` here. In any case, the runtime always uses `BinaryStringData` anyway under the hood, so even an unchecked cast should be fine.

Reply:
Great! Haven't looked into that at all. Thx @slinkydeveloper !